### ISSUE TITLE
ArviZ related experiments

### ISSUE THREAD

**Issue by OriolAbril** [smoh/kinesis#2]
Hi,

I am a participant of the Google Summer of Code with ArviZ, and I recently finished rewriting and extending the diagnostics and model comparison and checking functionalities. Now, as the last part of my project, I am trying out this new functions in real use cases. I saw this repository in ArviZ's [dependencies page](https://github.com/arviz-devs/arviz/network/dependents) while searching for suitable examples and I saw that the model uses multivariate normal variables.  Due to their shapes of the data and its pointwise likelihood they are very interesting to test for implementation bugs I may have missed in the CI tests. I hope you don't mind I experiment on my fork. 

I am creating this issue basically to inform you about it, so that you may check my changes and (if you have time and energy) maybe comment on them to see if they would be useful at all for you. Also, I can send some pull requests for some minor things (i.e. I had some troubles installing the package that I managed to solve modifying slightly `__init__.py` and `setup.py` if it is of your interest I can create a PR).

For now, I have just started using ArviZ everywhere possible on the [example notebook](https://github.com/OriolAbril/kinesis/blob/arviz/notebooks/model_validation.ipynb) (I think it looks quite good in general, but I am not convinced with comparing models with `plot_posterior`), and I will implement the new diagnostics and stats functions in the following days.

Thanks for everything.

**Comment by smoh** [comment]
Hi,

Thanks for letting me know. Yes, I'm fine with you using the code for arviz. I'll check out the fork occasionally (the arviz branche, right?).

> Also, I can send some pull requests for some minor things (i.e. I had some troubles installing the package that I managed to solve modifying slightly __init__.py and setup.py if it is of your interest I can create a PR).

Please do, that'd be helpful.

**Comment by OriolAbril** [comment]
Yes, it is the arviz branch, I will make it the default of my fork.

> Please do, that'd be helpful.

I will send a PR in the following days with this installation issue then.

---

Also, more on a modelling note, would you mind explaining and pointing out some references on why only the yy and zz components of C are corrected when defining D? I am asking mainly from an statistics point of view. If you look at the marginal posterior predictive checks, you can see that the x component in the model is not distributed like the observed data (whereas the other two look very similar):

![image](https://user-images.githubusercontent.com/23738400/62877161-c26f6d00-bd26-11e9-9e8b-22dd0d61c473.png)

There are more details on the notebook.

As an example, I have run [one model](https://github.com/OriolAbril/kinesis/blob/loo-pit-example/notebooks/model_validation.ipynb) without correcting the yy component, which yield the following result, where now both x and y components are miss-calibrated:

![image](https://user-images.githubusercontent.com/23738400/62877369-27c35e00-bd27-11e9-8380-1bc668fe46ab.png)

In these examples, the first row does not give much relevant info, as the min is equal to the z component and the max equal to the y component, you can omit it, I just want to test many cases.

**Comment by smoh** [comment]
Hi, Sorry for my late reply.

I have looked at the `generated_quantities` block and your notebook but cannot offer a clear answer to the figure. It would make more sense to me if posterior predictive samples of `a[0]` were under-dispersed than the observed...

> Also, more on a modelling note, would you mind explaining and pointing out some references on why only the yy and zz components of C are corrected when defining D?

It's because `sigv` parameter only affects the last two components of paramater `a`. You may see  [Lindegren et al. 2000](https://ui.adsabs.harvard.edu/abs/2000A%26A...356.1119L/abstract) for the model.

I'd like to try your notebook but I get the following error when initializing azfit -- do you have any idea why? Am I to use develop version of arviz or something like that?

```python
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-11-ad0d93baf52b> in <module>
      8     "observed_data": ["a", "rv"]
      9 }
---> 10 azfit = az.from_pystan(stanfit, **azkwargs)

~/.conda/envs/matrix/lib/python3.7/site-packages/arviz/data/io_pystan.py in from_pystan(posterior, posterior_predictive, prior, prior_predictive, observed_data, log_likelihood, coords, dims, posterior_model, prior_model)
    607             log_likelihood=log_likelihood,
    608             coords=coords,
--> 609             dims=dims,
    610         ).to_inference_data()

~/.conda/envs/matrix/lib/python3.7/site-packages/arviz/data/io_pystan.py in to_inference_data(self)
    152         return InferenceData(
    153             **{
--> 154                 "posterior": self.posterior_to_xarray(),
    155                 "sample_stats": self.sample_stats_to_xarray(),
    156                 "posterior_predictive": self.posterior_predictive_to_xarray(),

~/.conda/envs/matrix/lib/python3.7/site-packages/arviz/data/base.py in wrapped(cls, *args, **kwargs)
     23                 if getattr(cls, prop) is None:
     24                     return None
---> 25             return func(cls, *args, **kwargs)
     26 
     27         return wrapped

~/.conda/envs/matrix/lib/python3.7/site-packages/arviz/data/io_pystan.py in posterior_to_xarray(self)
     57         ignore = posterior_predictive + log_likelihood + ["lp__"]
     58 
---> 59         data = get_draws(posterior, ignore=ignore)
     60 
     61         return dict_to_dataset(data, library=self.pystan, coords=self.coords, dims=self.dims)

~/.conda/envs/matrix/lib/python3.7/site-packages/arviz/data/io_pystan.py in get_draws(fit, variables, ignore)
    403             for key, loc in keys_locs:
    404                 ary_slice = tuple(axes + loc)
--> 405                 ary[ary_slice] = pyholder.chains[key][-ndraw:]
    406         data[var] = ary
    407 

IndexError: index 3 is out of bounds for axis 2 with size 3
```

**Comment by smoh** [comment]
Ah, never mind -- I think it's just because I didn't pull your changes in models.py to retrieve more params. I'll let you know of any update.

**Comment by OriolAbril** [comment]
At least for now you'll need the development version of ArviZ as many functions have just been added during the summer.

I have given a quick read to the paper, thanks a lot. I will correct the notation whenever I have time, I was too fast to assume that three components were cartesian coordinates. About the overdispersion of the parallax data, I will check for typos in my notebook just in case, otherwise, could there be a typo in some unit conversion when defining the observed data or something like this? Then again, my understanding of both kinesis and the astrophysics behind the model is quite limited and my knowledge on gapipes is null.

Thanks again for all the explanations and attention.

**Comment by smoh** [comment]
Yes, I noticed and installed dev version of arviz.

I have not yet fully grasped how loo-pit is calculated (some of the references I could find like https://arxiv.org/pdf/1709.01449.pdf wasn't quite sufficient), but I think this is indeed related to the fact that the first dimension of vector `a` is not part of the Gaussian component (it does not make sense to flatten along 'star' axis for a[0]), and my intuition is that the reason it is aggregating at 0.5 for a[0] (the x component) is becasue it is in fact mean of uniform (which loo-pit should display if nothing is wrong).

I couldn't figure out how to calculate this just along sample axis treating it as a scalar parameter: `plot_loo_pit` seems to lack selecting `coords` or `var_names` like other functions in arvix. But if I compare the posterior predictive a[0] for i-th star directly, it is as I expect peaked around observed [0] for i-th star. As I said, I am not modelling a[0] hierarchically just the other two dimensions so flattening it along star axis is not particularly relevant.

> could there be a typo in some unit conversion when defining the observed data or something like this?

I am checking it against mock data generated pretty much independently from the stan part of things and quite confident there isn't any unit conversion error.

Hope what I said makes sense. Since your goals is to test arviz stuff, you may want to just write simpler multivariate normal model.

**Comment by OriolAbril** [comment]
Regarding LOO-PIT, I know there are not many references about it, which is why I also included "regular" posterior predictive checks. It is still a work in progress but I have a jupyter notwbook about LOO-PIT algorithm and its interpretation: https://github.com/OriolAbril/oriolabril.github.io/blob/loo-pit/notebooks/loo-pit/loo-pit-tutorial.ipynb. It may help. 

Regarding the previous comment about the parallax data, there are various things to take into account. First is that LOO-PIT is only for scalar quantities, which iswhy there is one plot per component, they are each treated as a scalar parameter. Secondly, what both LOO-PIT and ppc checks are telling is that if we do not flatten along star dim, the observed parallax of each star will be exactly the mean of the posterior predictive pdf, which if they both actually followed the same distribution, it is not what would happen. Hence my first question.

**Comment by smoh** [comment]
Thanks for the reference.

> Regarding the previous comment about the parallax data, there are various things to take into account. First is that LOO-PIT is only for scalar quantities, which iswhy there is one plot per component, they are each treated as a scalar parameter.

Yes, I understand LOO-PIT is for scalar quantities, yet the `a` parameter in my model is defined per star as in there are xyz for each star but the plots are not being made per i-th star, which is why I'm confused.

> Secondly, what both LOO-PIT and ppc checks are telling is that if we do not flatten along star dim, the observed parallax of each star will be exactly the mean of the posterior predictive pdf, which if they both actually followed the same distribution, it is not what would happen. Hence my first question.

I'm thoroughly confused and not following what is the issue here -- I think I expect that the mean of i-th star's posterior predictive distribution of a[0] (parallax) to be similar to i-th star's observed a[0] (parallax) because that's what I specify in my model. You're saying that's not what it should be?

**Comment by OriolAbril** [comment]
> I'm thoroughly confused and not following what is the issue here -- I think I expect that the mean of i-th star's posterior predictive distribution of a[0] (parallax) to be similar to i-th star's observed a[0] (parallax) because that's what I specify in my model. 

Yes indeed, however, distributions do not only have a mean but also a variance. I'll try to use one 1D example to make myself clear. If I had 100 observations of a random variable that I model as a N(0,1), you would expect most of the observations to be around 0, however, having std=1, around 30 observations would be outside the interval (-1,1). I all observations were to be between -0.1 and 0.1, then it would probably mean that the std of the normal variable is around 0.1 or smaller, not 1 as we assumed.

In your case it is a little trickier because each star has its own mean and std, however, the same still applies. It strikes as odd that of more than one hundred observations all of them are so close to its corresponding posterior predictive mean.
